File: src/main/java/com/philhanna/flashcards/domain/StudySession.java
package com.philhanna.flashcards.domain;

import com.philhanna.flashcards.domain.session.Session;
import com.philhanna.flashcards.domain.session.SessionCard;
import com.philhanna.flashcards.port.in.StudySessionUseCase;

/**
 * Application service that implements the {@link StudySessionUseCase} inbound
 * port. Coordinates domain objects ({@link Session}, {@link SessionCard}) to
 * carry out the core study-session use cases, keeping all domain logic out of
 * the UI adapter.
 */
public class StudySession implements StudySessionUseCase {

   private final Session session;

   public StudySession(Session session) {
      this.session = session;
   }

   @Override
   public void markRight() {
      SessionCard card = session.getCurrentCard();
      if (card != null)
         card.markRight();
      session.nextCard();
   }

   @Override
   public void markWrong() {
      SessionCard card = session.getCurrentCard();
      if (card != null) {
         card.markWrong();
         session.rotate();
         session.shuffle();
      }
   }

   @Override
   public void nextCard() {
      session.nextCard();
   }

   @Override
   public void previousCard() {
      session.previousCard();
   }

   @Override
   public Session getSession() {
      return session;
   }
}

File: src/flashcards/usecase/study_session.py
# flashcards.usecase.study_session

from flashcards.domain.session.session import Session
from flashcards.port.study_session_use_case import StudySessionUseCase


class StudySession(StudySessionUseCase):

    def __init__(self, session: Session) -> None:
        self._session = session

    def mark_right(self) -> None:
        card = self._session.current_card
        if card is not None:
            card.mark_right()
        self._session.next_card()

    def mark_wrong(self) -> None:
        card = self._session.current_card
        if card is not None:
            card.mark_wrong()
            self._session.rotate()
            self._session.shuffle()

    def next_card(self) -> None:
        self._session.next_card()

    def previous_card(self) -> None:
        self._session.previous_card()

    @property
    def session(self) -> Session:
        return self._session
